File: src/704.Binary Search/704.Binary Search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        while left <= right:
            middle = (left + right) // 2

            if nums[middle] < target:
                left = middle + 1
            elif nums[middle] > target:
                right = middle - 1
            else:
                return middle
        return -1


if __name__ == "__main__":
    list = [-1, 0, 3, 5, 9, 12]
    target = 9
    print(Solution().search(list, target))

File: src/704.Binary Search/704.Binary Search.go
package main

import "fmt"

func search(nums []int, target int) int {
	left := 0
	right := len(nums) - 1
	for left <= right {
		mid := left + (right-left)/2 // 防止直接相加造成溢出
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			left = mid + 1 // target 在右区间[mid + 1, right]
		} else {
			right = mid - 1 // target 在左区间[left, mid - 1]
		}
	}
	return -1
}

func main() {
	nums := []int{-1, 0, 3, 5, 9, 12}
	target := 9
	index := search(nums, target)
	fmt.Println(index)
}
